packages/image-cropper/ImageCropper.tsx

```diff
@@ -0,0 +1,458 @@
+import clsx from 'clsx';
+import normalizeWheel from 'normalize-wheel';
+import React from 'react';
+
+import {
+  computeCroppedArea,
+  getDistanceBetweenPoints,
+  getMidpoint,
+  restrictPosition,
+  restrictValue
+} from './cropUtils';
+import type { Area, MediaSize, Point, Size } from './types';
+
+export type CropperProps = {
+  image?: string;
+  cropSize: Size;
+  cropPosition: Point;
+  borderSize: number;
+  zoom: number;
+  zoomSpeed: number;
+  onCropChange: (location: Point) => void;
+  onZoomChange?: (zoom: number, maxZoom: number) => void;
+  onCropComplete?: (croppedAreaPixels: Area) => void;
+};
+
+type State = {
+  hasWheelJustStarted: boolean;
+};
+
+type GestureEvent = UIEvent & {
+  scale: number;
+  clientX: number;
+  clientY: number;
+};
+
+class ImageCropper extends React.Component<CropperProps, State> {
+  static defaultProps = {
+    zoom: 1,
+    zoomSpeed: 1
+  };
+
+  imageRef: React.RefObject<HTMLImageElement> = React.createRef();
+  containerRef: HTMLDivElement | null = null;
+  styleRef: HTMLStyleElement | null = null;
+  mediaSize: MediaSize = { width: 0, height: 0, naturalWidth: 0, naturalHeight: 0 };
+  maxZoom: number | null = null;
+  dragStartPosition: Point = { x: 0, y: 0 };
+  dragStartCrop: Point = { x: 0, y: 0 };
+  gestureZoomStart = 0;
+  isTouching = false;
+  lastPinchDistance = 0;
+  rafDragTimeout: number | null = null;
+  rafPinchTimeout: number | null = null;
+  wheelTimer: number | null = null;
+  currentDoc: Document = document;
+  currentWindow: Window = window;
+
+  state: State = {
+    hasWheelJustStarted: false
+  };
+
+  componentDidMount() {
```

Comment:
is it possible to migrate to functional component?

Reply:
I tried it before, and I think there were some problems, or seemed like a lot of work. But I don't remember exactly, so I'll look at it again.

Reply:
Cool not urgent